Riot/Modules/Room/Views/BubbleCells/Call/Group/RoomGroupCallStatusBubbleCell.swift

```diff
@@ -0,0 +1,334 @@
+            return
+        }
+        
+        let events = bubbleCellData.allLinkedEvents()
+        
+        NSLog("[RoomGroupCallStatusBubbleCell] render: \(events.count) events: \(events)")
+        
+        guard let widgetEvent = events
+                .first(where: {
+                    $0.eventType == .custom &&
+                        ($0.type == kWidgetMatrixEventTypeString || $0.type == kWidgetModularEventTypeString)
+                }) else {
+            return
+        }
+        
+        guard let widgetId = widgetEvent.stateKey else {
+            return
+        }
+        
+        guard let room = bubbleCellData.mxSession.room(withRoomId: widgetEvent.roomId) else {
+            return
+        }
+        
+        callDurationString = readableCallDuration(from: widgetEvent, endEvent: nil)
+        isIncoming = widgetEvent.sender != bubbleCellData.mxSession.myUserId
+        self.widgetEvent = widgetEvent
+        self.widgetId = widgetId
+        innerContentView.callIconView.image = Asset.Images.callVideoIcon.image
+        innerContentView.callTypeLabel.text = VectorL10n.eventFormatterCallVideo
+        
+        if isIncoming && !isJoined &&
+            TimeInterval(widgetEvent.age)/MSEC_PER_SEC < Constants.secondsToDisplayAnswerDeclineOptions {
+            
+            if JitsiService.shared.isWidgetDeclined(withId: widgetId) {
+                innerContentView.callerNameLabel.text = room.summary.displayname
+                room.summary.setRoomAvatarImageIn(innerContentView.avatarImageView)
+                
+                viewState = .declined
+                statusText = VectorL10n.eventFormatterCallYouDeclined
+            } else {
+                innerContentView.callerNameLabel.text = VectorL10n.eventFormatterGroupCallIncoming(bubbleCellData.senderDisplayName, room.summary.displayname)
+                
+                innerContentView.avatarImageView.setImageURI(bubbleCellData.senderAvatarUrl,
+                                            withType: nil,
+                                            andImageOrientation: .up,
+                                            toFitViewSize: innerContentView.avatarImageView.frame.size,
+                                            with: MXThumbnailingMethodCrop,
+                                            previewImage: bubbleCellData.senderAvatarPlaceholder,
+                                            mediaManager: bubbleCellData.mxSession.mediaManager)
+                
+                viewState = .ringing
+                statusText = nil
+            }
+        } else {
+            innerContentView.callerNameLabel.text = room.summary.displayname
+            
+            room.summary.setRoomAvatarImageIn(innerContentView.avatarImageView)
+        }
+        
+        innerContentView.avatarImageView.defaultBackgroundColor = .clear
+        
+        room.state { [weak self] (roomState) in
```

Comment:
Probably more optimized to make data management outside the cell and make specific reusable cell for each state.

Reply:
Ftr, i'll refactor this in future.